### ISSUE TITLE
Crashes immediately after "Read Flash" selected

### ISSUE THREAD

**Issue by dwildstr** [tuya-cloudcutter/cloudcutter-android#1]
I was trying to pull the firmware on a Daybetter RGBCW bulb (which seems to have had an update to PSK 02, so that the established profile no longer works) with release 0.40 and got through most of the process successfully. The screen with "Read Flash", "Read Storage", etc. came up, and when I hit "Read Flash" the app crashed. This happened pretty reliably, so I did it while running logcat to see what was causing it. Here are the last few entries:
```12-06 08:30:41.537 22979 22979 D LightleakService: Sending: io.github.cloudcutter.work.service.lightleak.command.FlashReadCommand@ea1adb2
12-06 08:30:41.539 22979 22979 D LightleakService: Awaiting response: io.github.cloudcutter.work.service.lightleak.command.FlashReadCommand@ea1adb2
12-06 08:30:41.539 22979 23451 D LightleakService: Running command: io.github.cloudcutter.work.service.lightleak.command.FlashReadCommand@ea1adb2
12-06 08:30:41.564 22979 23451 D LightleakService: Reading data #0, offset=0x0, count=8
12-06 08:30:41.589 22979 23451 E AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-5
12-06 08:30:41.589 22979 23451 E AndroidRuntime: Process: io.github.cloudcutter, PID: 22979
12-06 08:30:41.589 22979 23451 E AndroidRuntime: java.net.ConnectException: Network is unreachable
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at sun.nio.ch.Net.connect0(Native Method)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at sun.nio.ch.Net.connect(Net.java:466)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at sun.nio.ch.DatagramChannelImpl.connect(DatagramChannelImpl.java:771)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at io.ktor.network.sockets.UDPSocketBuilderJvmKt.connectUDP(UDPSocketBuilderJvm.kt:23)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at io.ktor.network.sockets.UDPSocketBuilder.connect(UDPSocketBuilder.kt:27)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at io.github.cloudcutter.work.protocol.UtilsKt.send(Utils.kt:44)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at io.github.cloudcutter.work.service.lightleak.LightleakService.flashRead(LightleakService.kt:173)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at io.github.cloudcutter.work.service.lightleak.LightleakService.access$flashRead(LightleakService.kt:40)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at io.github.cloudcutter.work.service.lightleak.LightleakService$onCommand$1.invokeSuspend(LightleakService.kt:127)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
12-06 08:30:41.589 22979 23451 E AndroidRuntime: 	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@a687bf1, Dispatchers.IO]
12-06 08:30:41.648 22979 23451 I Process : Sending signal. PID: 22979 SIG: 9
```

This is on a Moto X4 phone running a nightly build of LineageOS 19 (Android 12).

I know this is very much experimental software and crashes are to be expected, but I'm really hoping there's a fix/workwaround, since it looks like I'm really close to success here.

**Comment by kuba2k2** [comment]
The error here means the network is disconnected - which simply means the AP mode exited, or the device crashed.

1. Which profile are you using? 
2. Does the process complete successfully when you choose "Unconfigured" state? Did you use another state to open the reading screen?
3. Does the app crash instantly when pressing Read Flash? Or does it take, say, 10 seconds?
4. Does the device stay in AP mode? (bulb blinking slowly)
5. If not 4), does the device reboot/crash instantly and lights up again? Or does it freeze and reboot after ~60 seconds?
6. Less important, but what kind of device are you using for the CustomAP?

**Comment by dwildstr** [comment]
Thanks for the response! I appreciate this project and hope that this ends up being an easy problem to work around.

1: I'm using the "Lightleak - BK7231N -Variant 1 (XOR)" profile. This bulb is certainly running on a BK7231N. I used the Variant JTAG profile earlier and had the same result, but I didn't get a logcat of that.
2: The crash I'm describing happens when I start from "Unconfigured". About 1/3 of the time during the "Check if device is exploitable" step, the process fails with the "The device doesn't respond to ping requests" failure message. When this happens, the process can be restarted (apparently successfully up until it crashes at the same point as the Unconfigured Start) from the state "Configured to join CustomAP".
3: It's immediate.
4: Yes.
6: A spare ESP32 I had sitting around.

**Comment by kuba2k2** [comment]
1. The JTAG profile is just for debugging, so it shouldn't be used anyway.
2. Did you manage to get from Unconfigured to the reading screen?
3. Did you remember to disconnect the device from Tuya app? (disconnect & wipe data)
4. Precisely when it gets to `Check if device is exploitable`, does the bulb stop blinking for any time? Does it start blinking faster and continue like that?
5. When it does `Check if device is exploitable` does it disconnect you from the device's WiFi? (look at the Android status bar)
6. Do you see logs of the ESP32? When app says it's configuring to join custom AP, does the ESP say that a station has connected?

**Comment by dwildstr** [comment]
Sorry not to get responses back yesterday --- spent the whole day in the office and had no time to come and try to break lightbulbs.

2. Sometimes; as I said, I get a failure to respond to pings on the "Check if device is exploitable" step about a third of the time.
3. Yes.
4. The bulb goes solid-on (not blinking at all) for the "Wait for device to stop responding" step, until the reboot at the "Connect to smart device wifi" step, when it goes back to slow-blink. Except for that section, the bulb is on slow-blink the whole time, never a fast-blink.
5. It does seem to be disconnected, and remains disconnected through the "Read Flash" screen. That seems maybe like an important thing --- it should be connected to the SmartLife-XXXX network still, shouldn't it?
6. It says "Station connected! Waiting 8000ms to stop custom AP", then "Timeout" a little later (8 seconds, presumably) at that time followed by "Starting default AP - SSID: LightleakIdle / Password: cl0udcutt3r!@#".

An interesting new development: on the runthrough I did to get the information above, I did not have an immediate crash on the "Read Flash" stage --- there was 45 seconds or so of the circular "processing" swirl on the Android device screen, followed by a crash. That unfortunately happened during a run when I was not watching the logcat and I haven't yet succeeded in reproducing it.

**Comment by kuba2k2** [comment]
4. That looks totally fine. Another question, does the bulb *ever* exit AP mode (slow blinking mode)? If you run the whole Lightleak process, and wait about 3 minutes, does it then exit AP mode?
5. Yes, it should be connected the whole time, starting at the second `Connect to smart device wifi` step, when you re-reboot it to AP mode. Try connecting to the network manually, to see if your phone kicks you out of it - it might detect a "non-internet" network and maybe it doesn't like it. Android usually shows a message, saying something like "do you want to stay connected".
6. If it really did connect, you can try repeating the process at "Configured to join CustomAP" state.
---
2. It might just happen that this particular firmware is actually not exploitable using Lightleak. It would really help if you could pair it to SmartLife/TuyaSmart and check the firmware version (in fact, I should've started by saying that).

**Comment by dwildstr** [comment]
4. AFAICT, it exits AP mode a few minutes after the end of the Lightleak process crashes. Power-cycling the light even once will put it back in AP mode though.
5. OK, I tried manually forcing a connect to the SmartLife network. It then has several seconds of the circular 'processing' swirl on the screen, and then crashes with the same "Network is unreachable" logcat error as before. While it's swirling, the logcat is spamming the following repeated message:
```
12-08 13:59:33.665 29359 29382 D LightleakService: Reading data #0, offset=0x0, count=8
12-08 13:59:33.683 29359 29382 D LightleakService: Reading data #0, offset=0x0, count=8
```

I'm thinking that somehow it's disconnected through the entirety of the process from "Check if device is exploitable" to the "Read Flash" option, which means that all that preparation it does for reading the flash isn't actually making it through to the device? But no step in the Lightleak process is failing, which I'd think not being able to communicate with the bulb would do.

Re firmware version: that I know. It's on 2.9.6. Out-of-the-box these bulbs had an old firmware that was flashable with a 'generic WB2L' profile in cloudcutter, but these ones that I had been using around the house had a silent firmware upgrade and no longer work with that profile (it has an SSL error).

**Comment by kuba2k2** [comment]
> I'm thinking that somehow it's disconnected through the entirety of the process from "Check if device is exploitable" to the "Read Flash" option, which means that all that preparation it does for reading the flash isn't actually making it through to the device? 

Yes, it needs to send all packets to configure the device properly.

> But no step in the Lightleak process is failing, which I'd think not being able to communicate with the bulb would do.

That's because these are UDP packets, and no response is expected. That's why I added these "ping" checks (check if device still responds)... to check if the device didn't crash.

What's weird is that the device doesn't seem to crash at all, from what you're saying. That would mean it's either being configured properly, or no packets are ever reaching it. Can you connect to the network manually (Lightleak process is not necessary here) and check the IP addresses your phone receives?

> Re firmware version: that I know. It's on 2.9.6. 

Just checked 2.9.6's firmware, it should be exploitable. Since we have the firmware dump... you can always just use cloudcutter with 2.9.6 profile and flash whatever you want. But it'd be cool if we managed to get LL working on that as well. So don't give up just yet :slightly_smiling_face:

**Comment by dwildstr** [comment]
Well, that certainly gives me a useful fallback plan, although I tried the cloudcutter run_flash and it bricked this bulb, so I think the newer desktop cloudcutter is having some sort of problem dealing with 7231Ns (I'd been using 7231T profiles with run_flash, which is probably not good), but I should probably raise that as an issue with cloudcutter proper. Anyways, that means I have to liberate another 2.9.6 bulb from elsewhere in the house before I can do the IP-address test (probably just as well to do so; eliminate the individual bulb as a source of potential error).  It's odd, that with a lot of bulbs with the exact same hardware in my house all connected to Tuya, about a third of them are on 2.9.6 firmware and the rest are on 1.2.16. The latter's easier to flash to OpenBeken, so I think I'll run experiments only on the ones that are already upgraded.

**Comment by Cossid** [comment]
> I tried the cloudcutter run_flash and it bricked this bulb, so I think the newer desktop cloudcutter is having some sort of problem dealing with 7231Ns (I'd been using 7231T profiles with run_flash, which is probably not good), but I should probably raise that as an issue with cloudcutter proper

I think you're mixing up terminology here.  If you use a T _profile_ on N or vice versa, it will simply fail to exploit.  I think what you've done here is flashed a T _app bin file_ on N or vice versa, of which, CloudCutter has no protection mechanism for, and I'm not sure if we can even add.  We can only validate that it's valid for Beken in general.  That is the only way you can actually potentially brick a device using CloudCutter.

**Comment by dwildstr** [comment]
Hmm. You are of course right, and it turns out that this might be related to my wondering, in my previous message, why among seemingly identical bulbs (came out of the same package and everything!), some updated to 2.9.6 and some didn't. Popped the globe off of my bricked bulb and saw "WB2L" on the chip. Previous bulbs from the same pack were CB2Ls. These do in fact use different versions of the 7231 chip, so the N firmware I'd been blithely flashing onto these bulbs didn't work on _this particular_ bulb. Guess I'd better look under the globe every time before I flash them (and also warn the forum I shared a teardown on about it).

**Comment by Cossid** [comment]
You can safely use CloudCutter to determine between T and N if you use run_detach first (provided there is an existing firmware match) before you flash if you are unsure.  When you select by firmware version/name, the name of the profile almost always has "bk7231t", "bk7231n", or "bk7231s" in it (where bk7231s actually bk7231t).  If you get an A- SSID, that means the profile you used works, and the chip associated with that profile is what is inside.  You won't be able to verify specific model of board (WB2L, WB3L, CB3L, etc), but you can distinguish between T and N.  N profiles do not work on T modules and vice versa.

After you successfully detach, you can still run run_flash on the bulb in the same way you normally would.

So far, all devices with firmware version 2.9.# are T devices

**Comment by kuba2k2** [comment]
You don't have to open it up each time you flash. Just remember to use T firmware with T profiles, and N firmware with N profiles.

**Comment by Cossid** [comment]
I have added an extra check to CloudCutter run_flash where if a chip type is specified in the filename of the selected firmware and it doesn't match the chip type in the selected profile, it will give a strong warning against continuing.  If there is no chip type in the filename, it will continue as normal, I couldn't find a clear byte content indicator for N vs T.  It's not fool-proof, but it's a good extra layer.

**Comment by kuba2k2** [comment]
Wait, I just noticed something (actually, @Cossid also did): you're using Lightleak BK7231N profile on 2.9.6 firmware, which is most likely BK7231T.

> Out-of-the-box these bulbs had an old firmware that was flashable with a 'generic WB2L' profile 

WB2L == BK7231T which matches your 2.9.6 firmware.

> about a third of them are on 2.9.6 firmware and the rest are on 1.2.16

1.2.16 == BK7231N == CB2L

**TL;DR: try Lightleak with BK7231T profile on these 2.9.6 bulbs.**

**Comment by dwildstr** [comment]
I've got good news. I'm an idiot. Lightleak works perfectly and the problem, as they say, exists between the keyboard and the chair. Ran it with a BK7231T profile and it downloaded flash painlessly (I'll post it in the appropriate issue soon, but I figured I'd report here first). In my defense, most of the bulbs in this package _were_ CB2Ls. Like, I flashed them with 7231N firmware and they worked like a dream. My certainty this specific one was also a CB2L/BK7231N was, although completely incorrect, not unreasonable.

Thank you all for your time, and my apologies for wasting it.
